### ISSUE TITLE
`rerun --dry-run` or similar

### ISSUE THREAD

**Issue by mih** [datalad/datalad#2510]
In the context of testing https://github.com/datalad/datalad-container it would be useful to be able to programatically  inspect `run` records. We already have `--script`, but a more uniform way would be to yield run records as results. `rerun` already does that, but there is no way to just get this info, but not execute at the same time.

The comments already hint towards a `--list-commands`, but `--dry-run` might be another approach, that just yields the records and leaves everything else to the generic machinery.

@kyleam what do you think?

**Comment by kyleam** [comment]
> The comments already hint towards a --list-commands

Yeah, each time I've looked at that I've thought, "meh, not really a priority because you can easily inspect things with git log --grep=RUNCMD, etc".  But ...

> [...] it would be useful to be able to programatically inspect run records

in that case, yeah, that seems more useful.  I haven't  given much thought to what information the results dict should actually include, though.

**Comment by mih** [comment]
Brain dump:

At the very minimum the entire run record, of course. In addition to that:
- `commit`: the shasum of the commit
- `timestamp`: of the commit
- `author`: of the commit
- `path`: dataset path

**Comment by mih** [comment]
Would would be really help full is an included changeset in the style of `datalad diff` (across repository boundaries). 

When using this to record updates in some kind of meaningful directory structure quite informative logs or visualizations can be built based on the corresponding JSON output.

**Comment by mih** [comment]
Annnd, if we can get structured data out, it would be super-useful to amend the run record with additional structured data!

**Comment by kyleam** [comment]
> Annnd, if we can get structured data out, it would be super-useful to amend the run record with additional structured data!

Sorry, I'm not following.  Can you give me an example?

(I'm working on this now, btw.)

**Comment by mih** [comment]
So, let's say I build this fancy pipeline system that knows a bit more why what kind of command was executed. It would want to record this information in the changeset somehow. E.g. 'results recomputed due to bug in software used for commit X.

We would not want to blow up the run record with all of that arbitrary info, but we also would not put it in the commit message as an unstructured English blurb.

A minimalistic solution would be if a system using `run` could say "add this one key/value combo to the run record".   It would then be able to point to a file that is produced as an output and recorded in the dataset. This file could have any format and size and only needs to be understood by said system. As a consequence, this system could use `rerun --dry-run` or better `rerun --report` to rediscoved the knowledge it deposited and build fancy stuff on top of it.

**Comment by kyleam** [comment]
OK, I think I've got it.

> using run could say "add this one key/value combo to the run record". It would then be able to point to a file that is produced as an output and recorded in the dataset. 

Right, I think that'd be useful.  (The same idea has come up in conversations with @yarikoptic.)

> As a consequence, this system could use rerun --dry-run or better rerun --report to rediscoved the knowledge it deposited and build fancy stuff on top of it.

OK, so there's nothing special for `rerun --dry-run` to do then because it will already include the run info dict.

**Comment by kyleam** [comment]
> What do you think about --report instead?

I think `--report` is a better name in terms of capturing the possible uses.  When I see `--dry-run`, I just think "run this first to make sure I don't screw anything up".

**Comment by kyleam** [comment]
Hmm, just tried to switch things over from `--dry-run` to `--report`, but the cli parser isn't cooperating:

```diff
diff --git a/datalad/interface/rerun.py b/datalad/interface/rerun.py
index 4c82d8fc1..3c9c6b63f 100644
--- a/datalad/interface/rerun.py
+++ b/datalad/interface/rerun.py
@@ -134,6 +134,10 @@ class Rerun(Interface):
             directory. If a dataset is given, the command will be
             executed in the root directory of this dataset.""",
             constraints=EnsureDataset() | EnsureNone()),
+        report=Parameter(
+            args=("--report",),
+            action="store_true",
+            doc="""todo"""),
         # TODO
         # --list-commands
         #   go through the history and report any recorded command. this info
@@ -150,8 +154,10 @@ def __call__(
             branch=None,
             message=None,
             onto=None,
-            script=None):
-
+            script=None,
+            report=False):
+        print(report)
+        assert 0
         ds = require_dataset(
             dataset, check_installed=True,
             purpose='rerunning a command')

```

```
% datalad rerun --report
usage: datalad [-l LEVEL] [--pbs-runner {condor}] [-C PATH] [--version]
               [--dbg] [--idbg] [-c KEY=VALUE]
               [-f {default,json,json_pp,tailored,'<template>'}]
               [--report-status {success,failure,ok,notneeded,impossible,error}]
               [--report-type {dataset,file}]
               [--on-failure {ignore,continue,stop}] [--cmd]
datalad: error: ambiguous option: --report could match --report-status, --report-type
```

**Comment by kyleam** [comment]
It's failing in the recently added 

https://github.com/datalad/datalad/blob/6a465867bea5b48c6ff416b3411202e96d45f16d/datalad/cmdline/main.py#L200-L201

@yarikoptic any ideas?

**Comment by yarikoptic** [comment]
am I blind or there is no PR yet?  thought to try/troubleshoot since on initial appearance I do not think it is due to my changes anyhow and not sure how it "failed" exactly

**Comment by kyleam** [comment]
> am I blind or there is no PR yet? thought to try/troubleshoot since on initial appearance I do not think it is due to my changes anyhow and not sure how it "failed" exactly

I'm showing the exact patch and failure one comment above.

**Comment by bpoldrack** [comment]
Re `--report`:

Shouldn't this be the very same issue as #2524 and therefore work now? Did you try after the fix was merged?

**Comment by kyleam** [comment]
> Re --report:
>
> Shouldn't this be the very same issue as #2524 and therefore work now?

No, it seems to be unrelated.

> Did you try after the fix was merged?

Yes, the base is a master that includes the fix.  FWIW it fails in the exact same way (1) if the base is a commit in between the culprit and the fix from #2524 or (2) if the base is before the culprit from #2524.

@yarikoptic has provided more details [here](https://github.com/datalad/datalad/pull/2544#issuecomment-391562408).

**Comment by bpoldrack** [comment]
> No, it seems to be unrelated.
> ...
> @yarikoptic has provided more details here.

Okay - missed that. Thanks.
